ledger/nomos-ledger/src/mantle/sdp/rewards/blend/current_session.rs

```diff
@@ -0,0 +1,133 @@
+use cryptarchia_engine::Epoch;
+use nomos_blend_message::reward::SessionRandomness;
+use nomos_blend_proofs::quota::inputs::prove::public::LeaderInputs;
+use nomos_core::sdp::ProviderId;
+use rpds::HashTrieMapSync;
+use zksign::PublicKey;
+
+use crate::{
+    EpochState,
+    mantle::sdp::{
+        SessionState,
+        rewards::blend::{RewardsParameters, target_session::TargetSessionState},
+    },
+};
+
+/// Immutable state of the current session.
+/// The current session is `s` if `s-1` is the target session for which rewards
+/// are being calculated.
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct CurrentSessionState {
+    /// Current session randomness
+    session_randomness: SessionRandomness,
+}
+
+impl CurrentSessionState {
+    const fn new(session_randomness: SessionRandomness) -> Self {
+        Self { session_randomness }
+    }
+
+    pub const fn session_randomness(&self) -> SessionRandomness {
+        self.session_randomness
+    }
+}
+
+/// Collects epoch states seen in the current session.
+/// The current session is `s` if `s-1` is the target session for which rewards
+/// are being calculated.
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct CurrentSessionTracker {
+    /// The latest epoch seen in the current session.
+    latest_epoch: Epoch,
+    /// Collecting leader inputs derived from epoch states seen in the current
+    /// session. These will be used to create proof verifiers after the next
+    /// session update.
+    leader_inputs: Vec<LeaderInputs>,
+}
+
+impl CurrentSessionTracker {
+    pub fn new(first_epoch_state: &EpochState, settings: &RewardsParameters) -> Self {
+        Self {
+            latest_epoch: first_epoch_state.epoch,
+            leader_inputs: vec![settings.leader_inputs(first_epoch_state)],
+        }
+    }
+
+    pub fn collect_epoch(&self, epoch_state: &EpochState, settings: &RewardsParameters) -> Self {
+        if epoch_state.epoch > self.latest_epoch {
```

Comment:
Do we want to throw a warning if the new epoch is not strictly `1` greater than the previous one?